src/csharp/Microsoft.Spark/Utils/UdfSerDe.cs

```diff
@@ -336,64 +333,10 @@ private static TypeData SerializeType(Type type)
         }
 
         private static Type DeserializeType(TypeData typeData) =>
-            s_typeCache.GetOrAdd(typeData,
-                td => LoadAssembly(td.AssemblyName, td.ManifestModuleName).GetType(td.Name));
-
-        /// <summary>
-        /// Return the cached assembly, otherwise attempt to load and cache the assembly
-        /// in the following order:
-        /// 1) Search the assemblies loaded in the current app domain.
-        /// 2) Load the assembly from disk using manifestModuleName.
-        /// </summary>
-        /// <param name="assemblyName">The full name of the assembly</param>
-        /// <param name="manifestModuleName">Name of the module that contains the assembly</param>
-        /// <returns>Cached or Loaded Assembly</returns>
-        private static Assembly LoadAssembly(string assemblyName, string manifestModuleName)
-        {
-            return s_assemblyCache.GetOrAdd(
-                assemblyName,
-                _ => new Lazy<Assembly>(
-                    () =>
-                    {
-                        foreach (Assembly asm in AppDomain.CurrentDomain.GetAssemblies())
-                        {
-                            if (asm.FullName.Equals(assemblyName))
-                            {
-                                return asm;
-                            }
-                        }
-
-                        return LoadAssembly(manifestModuleName);
-                    })).Value;
-        }
-
-        /// <summary>
-        /// Returns the loaded assembly by probing the following locations in order:
-        /// 1) The working directory
-        /// 2) The directory of the application
-        /// If the assembly is not found in the above locations, the exception from
-        /// AssemblyLoader() will be propagated.
-        /// </summary>
-        /// <param name="manifestModuleName">The name of assembly to load</param>
-        /// <returns>The loaded assembly</returns>
-        private static Assembly LoadAssembly(string manifestModuleName)
-        {
-            string currDirAsmPath =
-                Path.Combine(Directory.GetCurrentDirectory(), manifestModuleName);
-            if (File.Exists(currDirAsmPath))
-            {
-                return AssemblyLoader(currDirAsmPath);
-            }
-
-            string currDomainBaseDirAsmPath =
-                Path.Combine(AppDomain.CurrentDomain.BaseDirectory, manifestModuleName);
-            if (File.Exists(currDomainBaseDirAsmPath))
-            {
-                return AssemblyLoader(currDomainBaseDirAsmPath);
-            }
-
-            throw new FileNotFoundException(
-                $"Assembly files not found: '{currDirAsmPath}', '{currDomainBaseDirAsmPath}'");
-        }
+            s_typeCache.GetOrAdd(
+                typeData,
+                td => AssemblyLoader.LoadAssembly(
+                    td.AssemblyName,
+                    td.ManifestModuleName).GetType(td.Name));
     }
 }
```

Comment:
Should we add some tests to ensure this behavior works and doesn't regress?

Reply:
We have some tests in UdfSerDeTests in Microsoft.Spark.UnitTest.  Do you think we need to add a different type of test there ?

Reply:
In my opinion, yes we need different tests. The reasoning being is that the current tests aren't failing with the current code in master. So they aren't testing the scenario you are trying to enable here.

Reply:
Added a few E2E tests.